### PR TITLE
Secure Join

### PR DESCRIPTION
**Pull request by mienaikoe**
- Coach can invite and uninvite people
- Moved auth-like endpoints from PersonController into SessionController
- Changed logic in Join to require an invite to be added to a team

TODO for myself: Manually Test Join after an invite

### CODE REVIEW

File: server/controllers/SessionController.js [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        findOne({ email }).
+        populate(DEEP_POPULATOR);
+
+      if( person && person.auth.passhash ){
+        throw new ValidationError("You've already joined.")
+      }
+
+      // Create if a Coach hasn't added this person yet
+      // Frontend needs to detect if team is null
+      if( !person ){
```
Comment (k3ntako) [inline server/controllers/SessionController.js]: Why do you create a person when they are not invited? Shouldn't this give an error?

I tried creating an account without inviting myself. It didn't log me in, but it created a session so I couldn't log in as anyone else.

File: server/controllers/SessionController.js [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+              $pull: { invites: person._id },
+              $push: { members: person._id }
+            })
+            // refresh person because its populated team is now outdated
+            person = await Person.
+              findOne({ _id: person._id }).
+              select(SAFE_SELECT).
+              populate(DEEP_POPULATOR);
+          } else {
+            console.warn("A Person with a set team is joining that was not marked as an invite to that team");
```
Comment (k3ntako) [inline server/controllers/SessionController.js]: I'm a bit confused with this warning.

what is a "set team"?

File: server/emails/team_invitation/text.pug [hunk 3]
```diff
@@ -0,0 +1,9 @@
+= "Hello,"
+
+= `Your coach, ${coach_name}, has invited you to join his team, ${team_name} on Wakebird.`
```
Comment (k3ntako) [inline server/emails/team_invitation/text.pug]: 1. i think you should have a comma after `${team_name}`.
2. you should change `his` to `their`

File: client/pages/Team/Roster.js [hunk 4]
```diff
@@ -24,20 +83,97 @@ export default function Roster(props) {
...
+      <h3> 
+        Members 
+        <div className={styles.actions}>
+          <Button
+            primary
+            raised
+            icon="person_add"
+            label="Add Member"
+            onClick={this.invitePersonModal}
+          />
```
Comment (k3ntako) [inline client/pages/Team/Roster.js]: This button should be visible only to coaches.

**Comment by mienaikoe** [comment]
Manual Test done and passed.

File: server/controllers/SessionController.js (hunk 2 above)
Comment (mienaikoe) [reply server/controllers/SessionController.js]: hmm guess I should .have said "A person with a team is joining".

Basically if `person.team` is not null

File: server/controllers/PersonController.js [hunk 5]
```diff
@@ -240,105 +143,91 @@ const PersonController = {
...
-      then ( (person) => {
-        if(!person){
-          throw new ForbiddenError("Email not found");
-        }
+      // validate team
+      const coachMatch = team.coaches.find( coach_id => coach_id.equals(req.session.person_id) )
+      if( !coachMatch ){
+        throw new ValidationError("You must be a coach to do this");
+      }
+      const memberMatch = team.members.find( member => member.email === email );
```
[the remark this answers is not in the document]
Comment (mienaikoe) [reply server/controllers/PersonController.js]: Yes. All coaches are members. Unless you think they shouldn't be?

[the remark this answers is not in the document]
Comment (mienaikoe) [reply server/controllers/PersonController.js]: Although that's a whole different conversation...

File: server/controllers/PersonController.js [hunk 6]
```diff
@@ -240,105 +143,91 @@ const PersonController = {
...
-        if(!req.body.token){
-          throw new ForbiddenError("No Password Reset Token");
-        }
-
-        let calculatedTokenHash = passhashify(req.body.token, person.auth.salt);
-
-        if(calculatedTokenHash !== person.auth.reset_token){
-          throw new ForbiddenError("Invalid Password Reset Token");
-        }
+  async uninviteFromMyTeam(req, res, next){
```
[the remark this answers is not in the document]
Comment (mienaikoe) [reply server/controllers/PersonController.js]: I mean we could, but then that would be a sad email to receive. I think it'd be better to just get the "no team" page and then ask the coach directly if there's a problem. 9/10 this would be caused by a typo.

File: server/controllers/SessionController.js (hunk 1 above)
Comment (mienaikoe) [reply server/controllers/SessionController.js]: Nah. We should still create the person, so the person doesn't have to do join again. This is for a case of like if a person joined Wakebird without getting an invite from their coach first.

File: client/pages/Team/Roster.js (hunk 4 above)
Comment (mienaikoe) [reply client/pages/Team/Roster.js]: This is fixed in the latest

File: server/controllers/PersonController.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (k3ntako) [reply server/controllers/PersonController.js]: I think they should be. I wanted to double check you didn't have to concat coaches.

File: server/controllers/SessionController.js (hunk 2 above)
Comment (k3ntako) [reply server/controllers/SessionController.js]: ahh okay, i was pretty tired when i was reading this. 

Why do you need a warning? Isn't this what we want?

File: server/controllers/SessionController.js [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+      }
+    }).then( person => {
+      let calculatedPasshash = passhashify(req.body.password, person.auth.salt)
+      if( calculatedPasshash !== person.auth.passhash ){
+        throw new ForbiddenError("Password Did Not Match");
+      }
+
+      // Session Creation
+      req.session.person_id = person._id
+      req.session.team_id = person.team._id;
```
Comment (k3ntako) [inline server/controllers/SessionController.js]: Pretty sure `person.team` is undefined, so `person.team._id` throws an error.

File: server/utilities/EmailUtils.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+const Email = require('email-templates');
+const path = require('path');
+const config = require('../config')[process.env.NODE_ENV];
+
+const transport = nodemailer.createTransport(config.mailer);
+const email = new Email({
+  message: {
+    from: `${process.env.GMAIL}@gmail.com`
+  },
+  transport: transport,
```
[the remark this answers is not in the document]
Comment (mienaikoe) [reply server/utilities/EmailUtils.js]: ok. How do we dictate when something is sent forreal vs not?

File: server/controllers/PersonController.js [hunk 9]
```diff
@@ -1,178 +1,90 @@
...
-      populate(FIND_POPULATOR).
+      populate(DEEP_POPULATOR).
       then( person => {
-        let isCoach = person.team.coaches.find( function(coach){
-          return coach._id.toString() === req.session.person_id.toString()
-        });
-        if( isCoach ){
-          return Person.findOne({ _id: req.params.person_id }).
-            select(COACH_SELECT);
+        if( !person ){
```
[the remark this answers is not in the document]
Comment (mienaikoe) [reply server/controllers/PersonController.js]: I think it'd be fine if a person tried to get `me` when they were only an invite. They wouldn't be able to login in the first place, so this would inevitably throw a 403 before the code even got here.

File: server/controllers/PersonController.js [hunk 10]
```diff
@@ -240,105 +143,91 @@ const PersonController = {
...
-      }).
-      then( () => {
-        res.json({okay: true});
-        next();
-      }).catch( next )
-  },
+      person = await Person.
+        findOne({ _id: person._id }).
+        select(COACH_SELECT).
+        populate(DEEP_POPULATOR);
```
[the remark this answers is not in the document]
Comment (mienaikoe) [reply server/controllers/PersonController.js]: true

[the remark this answers is not in the document]
Comment (mienaikoe) [reply server/controllers/PersonController.js]: actually hold up. `Person.findOne` does depend on `team.save` because it's populating the person with the person's team.

File: server/controllers/SessionController.js (hunk 2 above)
Comment (mienaikoe) [reply server/controllers/SessionController.js]: I'll reword this to be more clear

File: server/emails/team_invitation/text.pug (hunk 3 above)
Comment (mienaikoe) [reply server/emails/team_invitation/text.pug]: ah wow good catch

File: server/controllers/SessionController.js [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+      let person = await Person.
+        findOne({ email }).
+        populate(DEEP_POPULATOR);
+
+      if( person && person.auth.passhash ){
+        throw new ValidationError("You've already joined.")
+      }
+
+      // Create if a Coach hasn't added this person yet
+      // Frontend needs to detect if team is null
```
[the remark this answers is not in the document]
Comment (mienaikoe) [reply server/controllers/SessionController.js]: wha? Non-Members for custom emails aren't `Person`s in our code, are they?

File: server/utilities/EmailUtils.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (k3ntako) [reply server/utilities/EmailUtils.js]: As long as we are using nodemailer, we set it up so all mail sent in testing and development go to Ethereal Email.

It should be in `server/config.js`.

File: server/controllers/PersonController.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (k3ntako) [reply server/controllers/PersonController.js]: ahh good point!

File: server/controllers/SessionController.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (k3ntako) [reply server/controllers/SessionController.js]: You're right, they aren't `Person`s. If they are already included in an email draft as a `nonMember` and they join the team, they would receive the email twice. 

Guess this is super particular, so we can deal with it later?

Screen Shot 2019-04-01 at 12 49 58 PM

File: server/controllers/SessionController.js [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+              _id: team._id
+            },{
+              $pull: { invites: person._id },
+              $push: { members: person._id }
+            })
+            // refresh person because its populated team is now outdated
+            person = await Person.
+              findOne({ _id: person._id }).
+              select(SAFE_SELECT).
+              populate(DEEP_POPULATOR);
```
Comment (k3ntako) [inline server/controllers/SessionController.js]: Ran into another issue with people who join first and then get invited.

Once they get invited, they still can't see the events. This is because in `Event.listForMyTeam` in the `EventController`, `req.session.team_id` is undefined. User has to log out and then log in, but this isn't apparent to the user because it throws no errors.

Comment (mienaikoe) [reply server/controllers/SessionController.js]: Hmm. So the best we could do is update it on refresh by ensuring the team is set on the session when calling `me`. You would still have to refresh... I'll go that far but any farther would require some complicated socket stuff that just isn't worth it right now.

Comment (k3ntako) [reply server/controllers/SessionController.js]: Yeah I think refreshing is fine. The biggest issue is that it goes from "No Team" to "No Events Yet!". This makes it seem like everything is functioning, and it just looks like coach just hasn't added events yet.

File: server/controllers/SessionController.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (mienaikoe) [reply server/controllers/SessionController.js]: Added a ticket for Formalizing a member model . https://trello.com/c/Kmz0drH8/163-tech-debt-member-model

Will close for now. The plan for later should be to make non-members `Person` without adding them to the team.

File: server/utilities/EmailUtils.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (mienaikoe) [reply server/utilities/EmailUtils.js]: true. I see it now.
